sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/DefaultSynchronousMetricStorage.java

```diff
@@ -208,25 +232,79 @@ public MetricData collect(
+    // hence during collect(), when the map is at full capacity, we try to clear away unused
+    // aggregator handles, so on next recording cycle using this map, there will be room for newly
+    // recorded Attributes. This comes at the expanse of memory allocations. This can be avoided
+    // if the user chooses to increase the maxCardinality.
+    if (memoryMode == REUSABLE_DATA && reset) {
+      if (aggregatorHandles.size() >= maxCardinality) {
+        aggregatorHandles.forEach(
+            (attribute, handle) -> {
+              if (!handle.isValuesRecorded()) {
+                aggregatorHandles.remove(attribute);
+              }
+            });
+      }
+    }
+
     // Grab aggregated points.
-    List<T> points = new ArrayList<>(aggregatorHandles.size());
     aggregatorHandles.forEach(
         (attributes, handle) -> {
-          T point = handle.aggregateThenMaybeReset(start, epochNanos, attributes, reset);
+          T point = null;
           if (reset) {
-            // Return the aggregator to the pool.
-            aggregatorHandlePool.offer(handle);
+            if (memoryMode == IMMUTABLE_DATA) {
+              point = handle.aggregateThenMaybeReset(start, epochNanos, attributes, reset);
+              // Return the aggregator to the pool.
+              aggregatorHandlePool.offer(handle);
+            } else /* REUSABLE_DATA */ {
+              // Handles persists across collect() in REUSABLE_DATA hence we should
+              // filter unused handles
+              if (handle.isValuesRecorded()) {
+                point = handle.aggregateThenMaybeReset(start, epochNanos, attributes, reset);
+              }
+            }
+          } else /* No reset (CUMULATIVE temporality) */ {
+            point = handle.aggregateThenMaybeReset(start, epochNanos, attributes, reset);
           }
+
           if (point != null) {
             points.add(point);
           }
         });
 
-    // Trim pool down if needed. pool.size() will only exceed maxCardinality if new handles are
-    // created during collection.
-    int toDelete = aggregatorHandlePool.size() - (maxCardinality + 1);
-    for (int i = 0; i < toDelete; i++) {
-      aggregatorHandlePool.poll();
+    if (reset) {
+      if (memoryMode == IMMUTABLE_DATA) {
+        // Trim pool down if needed. pool.size() will only exceed maxCardinality if new handles are
+        // created during collection.
+        int toDelete = aggregatorHandlePool.size() - (maxCardinality + 1);
+        for (int i = 0; i < toDelete; i++) {
+          aggregatorHandlePool.poll();
+        }
+      } else /* REUSABLE_DATA */ {
+        // Pool is unused in REUSABLE_DATA since it allocated memory upon each use.
+        // Instead of keep the aggregator handles map which is not active,
+        // and use it in the beginning of next collect() to initialize the
+        // newly active aggregatorHandles map.
+        previousCollectionAggregatorHandles = aggregatorHandles;
+      }
     }
```

Comment:
```suggestion
    // Trim pool down if needed. pool.size() will only exceed maxCardinality if new handles are
    // created during collection.
    int toDelete = aggregatorHandlePool.size() - (maxCardinality + 1);
    for (int i = 0; i < toDelete; i++) {
      aggregatorHandlePool.poll();
    }

    if (reset && memoryMode == REUSABLE_DATA) {
      previousCollectionAggregatorHandles = aggregatorHandles;
    }
```
Let's keep trimming the pool in all conditions. Doing so acts as an additional safety check against runaway memory usage due to programming errors. Also, I left a comment above suggesting keeping returning handles to the pool when memory mode is reusable.

Reply:
Saw you respond to this comment [here](https://github.com/open-telemetry/opentelemetry-java/pull/5998#issuecomment-1824440844). Keeping the conversation in this thread so we can track / resolve it.

I think you're probably right, but maybe it was possible to for the pool to exceed the max cardinality at some point. I still think its a good safety check and its cheap to perform.

Reply:
Ok. I took your suggestion and made it trim always. 
The only thing is that I couldn't think of scenario for it hence the code coverage won't cover this section of the code as codecov indicated.